Devel

wjlei1990 opened the pull request.

wjlei1990 committed:
Add more debug output information and fix bugs

wjlei1990 committed:
change c_1(minimum window length parameter) to array

wjlei1990 committed:
Add methods that write plain text window output file

wjlei1990 committed:
Change function order(put global data quality check in the beginning)

wjlei1990 committed:
modify window selection strategy(selection region based on surface wave)

wjlei1990 committed:
Merge branch 'devel' of https://github.com/wjlei1990/pyflex into devel

wjlei1990 committed:
slightly modify curtail window method, data_quality_check return method and data_fit method

wjlei1990 commented:
Hi Lion,

I have some updates on pyflex. Currently, I am benchmarking flexwin and pyflex and I found there are a few differences  in implementation details. So I modified some based on my own judgements.

I have some personal changes(probably you may not like it...) and also some bug fix. Do you need me to provide you a detailed doc on every commits? If that works better for you, I can do it.

krischer commented:
Hi Wenjie,

> Do you need me to provide you a detailed doc on every commits? If that works better for you, I can do it.

Can you comment on the logic of some of these changes? Just do it here in this issue - that's fine.

> I found there are a few differences in implementation details

Yea there are some. Mostly on purpose - they are documented here:
http://krischer.github.io/pyflex/#difference-to-the-original-flexwin-algorithm

The goal of pyflex is not to be identical with FLEXWIN but improve upon it where possible. That said I am happy to take any chances that either improve the algorithm or the execution speed.

I'll comment on the code inline.

You'll also have to make sure that the tests pass and please also document all the changes in the docstring and the documentation.

krischer commented on src/pyflex/config.py:
```diff
@@ -264,6 +267,13 @@ def __init__(self, min_period, max_period, stalta_waterlevel=0.07,
 
+        if select_mode.lower() not in ["all_wave", "body_and_surface", "body_wave",
+                "surface_wave", "mantle_wave"]:
+            raise PyflexError(
```
Better raise a `ValueError`.

krischer commented on src/pyflex/config.py:
```diff
@@ -32,8 +33,9 @@ def __init__(self, min_period, max_period, stalta_waterlevel=0.07,
-                 resolution_strategy="interval_scheduling"):
+                 window_signal_to_noise_type="energy",
+                 resolution_strategy="interval_scheduling",
+                 select_mode="body_and_surface"):
```
also has to be documented in the docstring. The default value should probably be "all_wave" to not change the results for people updating.

krischer replied:
Forget that. I read the code in the meanwhile and `body_and_surface` is the right choice.

krischer commented on src/pyflex/config.py:
```diff
@@ -32,8 +33,9 @@ def __init__(self, min_period, max_period, stalta_waterlevel=0.07,
                  window_weight_fct=None,
-                 window_signal_to_noise_type="amplitude",
-                 resolution_strategy="interval_scheduling"):
+                 window_signal_to_noise_type="energy",
```
please leave the old default value

krischer commented on src/pyflex/config.py:
```diff
@@ -24,7 +24,8 @@ def __init__(self, min_period, max_period, stalta_waterlevel=0.07,
                  cc_acceptance_level=0.7, s2n_limit=1.5, earth_model="ak135",
-                 min_surface_wave_velocity=3.0,
+                 min_surface_wave_velocity=3.5,
+                 max_surface_wave_velocity=4.2,
```
This needs to be documented in the docstring of the config object.

krischer commented on src/pyflex/config.py:
```diff
@@ -24,7 +24,8 @@ def __init__(self, min_period, max_period, stalta_waterlevel=0.07,
                  dlna_acceptance_level=1.3, dlna_reference=0.0,
                  cc_acceptance_level=0.7, s2n_limit=1.5, earth_model="ak135",
-                 min_surface_wave_velocity=3.0,
+                 min_surface_wave_velocity=3.5,
```
please don't change existing default values. There are already people using this and we don't want their results to change just because they update pyflex.

krischer commented on src/pyflex/__init__.py:
```diff
@@ -43,6 +43,7 @@ class PyflexWarning(UserWarning):
 
 # Setup the logger.
 logger = logging.getLogger("pyflex")
+#logger.setLevel(logging.WARNING)
```
please remove the comment

krischer commented on src/pyflex/config.py:
```diff
@@ -264,6 +267,13 @@ def __init__(self, min_period, max_period, stalta_waterlevel=0.07,
         self.resolution_strategy = resolution_strategy.lower()
 
+        if select_mode.lower() not in ["all_wave", "body_and_surface", "body_wave",
+                "surface_wave", "mantle_wave"]:
```
I guess it makes sense to either add the `_wave` suffix to `body_and_surface` or remove it from all others. I would also rename the argument to `selection_mode`.

krischer commented on src/pyflex/config.py:
```diff
@@ -271,7 +281,7 @@ def _convert_to_array(self, npts):
         attributes = ("stalta_waterlevel", "tshift_acceptance_level",
                       "dlna_acceptance_level", "cc_acceptance_level",
-                      "s2n_limit")
+                      "s2n_limit", "c_1")
```
At that point it might be a good idea to turn all the `c_` parameters into arrays?

wjlei1990 replied:
Yes...Do you want to do it right now or after merge?

krischer replied:
That can well be done after the merge.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -184,6 +184,21 @@ def default(self, obj):
+        nw = self.observed.stats.network
+        comp = self.observed.stats.channel
+        loc = self.observed.stats.location
+        f = open(filename, 'w')
```
Please use a `with` statement to assure the file is closed in all circumstances.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -184,6 +184,21 @@ def default(self, obj):
+        comp = self.observed.stats.channel
+        loc = self.observed.stats.location
+        f = open(filename, 'w')
+        text = "%s.%s.%s.%s\n" %(sta, nw, comp, loc)
```
Can be replaced with 

``` python
text = "%s\n" % self.observed.id
```

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -184,6 +184,21 @@ def default(self, obj):
             except TypeError:
                 filename.write(j.encode())
 
+    def write_text(self, filename):
```
Whats the motivation for this function? I.e. when would I want to use this instead of the JSON serialization? I'm not against having it - just wundering.

This function also has to be added to the tests.

wjlei1990 replied:
I think the purpose is for those people who want to use the window output directly for other programs other than python. The plain txt file contains much less information than the json file, but it is more friendly to other programs like C and FORTRAN to read in.  What do you think?

I will add the test later :)

krischer replied:
Yea that makes some sense :-)

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -420,6 +452,8 @@ def reject_based_on_signal_to_noise_ratio(self):
         noise = self.observed.data[self.config.noise_start_index:
                                    self.config.noise_end_index]
 
+        offset = self.event.origin_time - self.observed.stats.starttime
```
please calculate that on the fly within the if clause where its needed.

wjlei1990 replied:
I think I am going to delete that...Cause I think it is not necessary to have this value. I am just going to set:
left = win.relative_starttime
right = win.relatvie_endtime

One more thing, for the method "reject_based_on_signal_to_noise_ratio", I feel like we should provide people with one more option that uses both criteria, "amplitude" and "energy". At least for my case, I would like to have both checked... People can also chose which mode they want to use based on the features of their data.

krischer replied:
> One more thing, for the method "reject_based_on_signal_to_noise_ratio", I feel like we should provide people with one more option that uses both criteria, "amplitude" and "energy". At least for my case, I would like to have both checked... People can also chose which mode they want to use based on the features of their data.

:+1: Could just be called `both` or `amplitude_and_energy` - I prefer the later one as its more descriptive.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -505,6 +543,45 @@ def calculate_ttimes(self):
+                    3 * self.config.min_period + offset
+        elif select_mode == "mantle_wave":
+            min_time = dist_in_km / self.config.max_surface_wave_velocity + offset
+            max_time = self.observed.stats.endtime - self.observed.stats.starttime
```
Its good practice to add an `else` clause and raise an error just in case something in a previous step goes wrong.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -688,10 +765,7 @@ def curtail_window_length(win):
-                (self.peaks != win.center)]
-            if len(internal_maxima) < 2:
-                return win
+                (self.peaks >= win.left) & (self.peaks <= win.right) ]
```
What is the motivation behind this? You don't want to add the current peak to the list of peaks to feed into the curtailing and it does not work with less then two additional peaks.

wjlei1990 replied:
I first compared yours with the original fortran version and I found the difference. I feel like the original fortran version makes more sense.  The curtailing left is just to curtail from the first left maxim. Same for the right maxim. It doesn't matter if the first left maxim is same as the center or the right maxium is same as the center. If in one window there is only one maxim(which is the center), it is still OK to curtail that window if the left or the right are too long. If you checked the FLEXWIN manual, in the curtail window example,  the example window just contains one maxim and it is curtailed.
What do you think?

krischer replied:
You are right. I misunderstood the curtailing a bit. Proceed! ;-)

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -406,6 +433,11 @@ def determine_signal_and_noise_indices(self):
+                    %(self.config.noise_start_index, self.config.noise_end_index))
+        logger.info("Signal start index and noise end index: %d, %d"
+                    %(self.config.signal_start_index, self.config.signal_end_index))
+
```
No need to log that as that is specified by the user in the configuration.

wjlei1990 replied:
But...en...People might leave those values as default value(None) in the configuration stage...So they may want to know the what the exact indexes are when pyflex is running...
What do you think?...

krischer replied:
True. I forgot about that. It might be nicer for users to read something in seconds, e.g.:

```
Noise from 0-113 seconds, signal from 113-1051 seconds.
```

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -316,25 +339,28 @@ def select_windows(self):
+        if self.config.check_global_data_quality:
+            if not self.check_data_quality():
+                return []
+
```
Good idea :-) 

You might have to a bit careful as signal and noise indices cannot be determined if either event or station information is missing. But the tests should catch that.

wjlei1990 replied:
Which part of the code are your referring to... :( aha

krischer replied:
It cannot determine the signal and noise onsets if it does not event or station information.

wjlei1990 replied:
I think I understood...

One thing, I feel like we need another check for noise_end_index(also for noise_start_index and other index) in case people assign unreasonable values to those index.

For example, for a 100 elements array. If people assign noise_start_index = 5 and noise_end_index = 15, It makes sense since noise_end_indx > noise_start_index. But if people set noise_start_idx = 5 and noise_end_index = 2, then it does not make sense. However, since python allows negative index, I think we need some tricks here to make sure it is robust. I am talking about extreme cases..but I think it is nice to have those kinds of checks.

If you already have those checks somewhere, just forget what I said...

krischer replied:
That would be good. We should probably just check that both indices are positive and the end index is larger then the start index. I think its reasonable to not allow negative indices although it would be valid in Python.

wjlei1990 replied:
I have written a method in config.py that convert negative index into positive index and then do the checking.

krischer replied:
I guess I'll wait for the implementation, but: How do you do that? The config object has no access to the data and thus cannot resolve negative indices...

wjlei1990 replied:
It is actually called in determin_signal_and_noise_indices(). At that time, all the indexes are determined and npts is already known.

```
def determine_signal_and_noise_indices(self):
        """
        Calculate the time range of the noise and the signal respectively if
        not yet specified by the user.
        """
        if self.config.noise_end_index is None:
            if not self.ttimes:
                logger.warning("Cannot calculate the end of the noise as "
                               "event and/or station information is not given "
                               "and thus the theoretical arrival times cannot "
                               "be calculated")
            else:
                self.config.noise_end_index = \
                    int(self.ttimes[0]["time"] - self.config.min_period)
        if self.config.signal_start_index is None:
            self.config.signal_start_index = self.config.noise_end_index

        self.config._convert_index(npts=self.observed.stats.npts)
```

krischer replied:
Yea that should work. I'd just noticed that we also need to copy the config object upon in initializing the window selector object. Otherwise we'll modify the config object and people might want to keep reusing the same object for multiple runs.

wjlei1990 replied:
Yes... That's what I am doing now...I just use copy.copy() to make a copy of config every time externally. That would be very helpful if we do it inside. 
I will leave this to you.

krischer replied:
Done in b3168622b064f96f25902cd4d575335da9f09e15.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -306,6 +321,14 @@ def calculate_preliminiaries(self):
             self.peaks = self.peaks[(self.peaks > min_trough) &
                                     (self.peaks < max_trough)]
 
+    def print_remaining_windows(self):
```
Can you make this a private function?

`__print_remaining_windows`

wjlei1990 replied:
Do you think it makes sense to put this function in window.py?

Since all it only requires window object.

krischer replied:
Yes I think that would make sense. Best just implement a `__str__` method for the window object. Then you can simply `print(self.windows)` for the same effect.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -232,12 +247,12 @@ def _parse_event_and_station(self):
-                tr = self.observed
-                ftype = "observed"
-            elif hasattr(self.synthetic.stats, "sac"):
+            if hasattr(self.synthetic.stats, "sac"):
```
Why did you invert that? I would tend to trust the observed data more than the synthetic sac headers.

wjlei1990 replied:
I think this is a personal choice then...

For my case, the observed data(sac format) usually doesn't contain the event information. So When you try to extract event information from the observed file, it will always fail. On the other hand, the synthetic is generated by specfem so the header information is more complete.

So what do you think?

krischer replied:
Your point is better :-)

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -505,6 +543,45 @@ def calculate_ttimes(self):
+        logger.info("Rejection based on travel times retained %i windows." %
+                    len(self.windows))
+
     def reject_on_traveltimes(self):
```
This function can probably be deleted - the `reject_on_select_mode()` method appears to be able to completely replace it.

wjlei1990 replied:
Agree. I am going to delete that.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -505,6 +543,45 @@ def calculate_ttimes(self):
         self.ttimes = sorted(tts, key=lambda x: x["time"])
         logger.info("Calculated travel times.")
 
+    def reject_on_select_mode(self):
```
This is really cool.

I think it could also be expanded to select windows for specific phase arrivals. TauP does already calculate the arrival times for various phases so this function could well filter windows that contain the chosen phases.

wjlei1990 replied:
Yes, Let's do that after merging?

krischer replied:
works for me.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -505,6 +543,45 @@ def calculate_ttimes(self):
+        elif select_mode == "body_wave":
+            min_time = self.ttimes[0]["time"] - self.config.min_period + offset
+            max_time = dist_in_km / self.config.max_surface_wave_velocity + \
+                    3 * self.config.min_period + offset
```
Why the factor of 3?

wjlei1990 replied:
Personal choice... haha :) I think people may get confuse when seeing this magic number, "3".
What would be the better solution? Put that config?

krischer replied:
I'm also not sure. I'm also fine with 3 but best use the same magic number for both - min and max times. I always use a single minimum period in an attempt to account for effects of potentially acausal filters.

I'd like to avoid having even more configuration parameters as the amount of configuration values is already a bit ridiculous and hard to understand ;-)

wjlei1990 replied:
One thing... Do we need to differentiate components? Since for Z and R component, the first arrival is P or Pn. But for T component, it is S or ScS. I think this will bring the code more complexity. I just want to leave a note and we can refine it later if needed...

Another thing, I just submit an update to FLEXWIN that windows that overlaps with the noise region should be rejected. In pyflex, we have the method --- " reject_on_selection_mode". So if we also also want to put the "reject_on_noise_region". There are two cases when the "reject_on_noise_region" will take effect:
1) the rejection_on_selection_mode doesn't take effect(missing station information or something) but windows overlapping with noise region should still be rejected anyway
2) if people somehow accidently [min_time, max_time](of reject_on_selection_mode) inside noise region, should pyflex raise an error or leave it out?

...One more thought. I think it is ok to keep both. We can state it in that way. [noise_start_index, noise_end_index] specify the noise region and [signal_start_index, signal_end_index] specify the signal region. But we have another control here, in signal region, we just want to select those windows we are interested in.

But in this case, the [signal_start_index, signal_end_index] becomes pretty weak now since it will only be used in the check_data_quality. What do you think if we put [signal_start, signal_end] = [min_time, max_time]?

krischer replied:
> One thing... Do we need to differentiate components? Since for Z and R component, the first arrival is P or Pn. But for T component, it is S or ScS. I think this will bring the code more complexity. I just want to leave a note and we can refine it later if needed...

In theory yes but I would just ignore this. pyflex does not care about components at all and it would not even have to be any of Z,N,E,R, or T. It just checks they are identical for both data and synthetics I believe. The additional complexity is not worth the trade-off IMHO.

> Another thing, I just submit an update to FLEXWIN that windows that overlaps with the noise region should be rejected. 

You want at least a bit of overlap to account for acausal filter effects - otherwise you'll never catch the first arrival. If you do that it should be good.

> In pyflex, we have the method --- " reject_on_selection_mode". So if we also also want to put the "reject_on_noise_region". There are two cases when the "reject_on_noise_region" will take effect:
> 1) the rejection_on_selection_mode doesn't take effect(missing station information or something) but windows overlapping with noise region should still be rejected anyway

I agree.

> 2) if people somehow accidently min_time, max_time inside noise region, should pyflex raise an error or leave it out?

Raise an error. I'm a big fan of raising early and forcing users to fix their mistakes.

A side note: These comments will kind of disappear once you push to this branch and change the line thats commented below. So before you push: can you copy all important comments into the comments section of this issue? It would be nice if they don't get lost!

wjlei1990 replied:
Could you tell me where is the "comments section of this issue" :P

I think as long as I keep this pull request and this branch(this commit in this branch, more precisely), this comment will not be deleted.

krischer replied:
;-)

The comment box at the end of this page: https://github.com/krischer/pyflex/pull/7

The comments will always remain whereas the one below a code line will disappear if the code line in question changed.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -721,9 +795,14 @@ def reject_windows_based_on_minimum_length(self):
         Reject windows smaller than the minimal window length.
         """
-        self.windows = list(filter(
-            lambda x: (x.right - x.left) >= self.minimum_window_length,
```
The minimum window length property can be deleted as it has become obsolete.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -808,8 +887,8 @@ def _sanity_checks(self):
-            warnings.warn("The amplitude difference between data and "
-                          "synthetic is fairly large.")
+            warnings.warn("The amplitude difference between data and"
+                           "synthetic is fairly large.")
```
can you revert this? The space in the first line is lost in the new version.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -745,18 +824,18 @@ def reject_based_on_criteria(win):
+            if not (tshift_min <= win.cc_shift *
+                    self.observed.stats.delta <= tshift_max):
+                logger.debug("Window rejected due to time shift:    %6.2f %6.1f %6.1f" %
+                             (win.cc_shift, win.relative_starttime, win.relative_endtime))
```
Can you make these a bit more descriptive? Otherwise it will be hard to make sense of them in a couple of months. Something along the lines of

``` python
logger.debug("Window [%.1f - %.1f seconds] rejected due to time shift: %.1f" % ...)
```

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -829,6 +908,7 @@ def plot(self, filename=None):
         # Lazy imports to not import matplotlib all the time.
         import matplotlib.pyplot as plt
         from matplotlib.patches import Rectangle
+        import numpy as np
```
numpy is already imported.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -865,9 +945,10 @@ def plot(self, filename=None):
+        obsd_handle, = plt.plot(times, self.observed.data, color="black")
+        synt_handle, = plt.plot(times, self.synthetic.data, color="red")
         plt.xlim(times[0], times[-1])
+        plt.legend([obsd_handle, synt_handle], ["obsd", "synt"], prop={'size':8})
```
Simpler:

``` python
plt.plot(times, self.observed.data, color="black", label="Observed")
plt.plot(times, self.synthetic.data, color="red", label="Synthetic")

...

plt.legend(prop={"size": 8})
```

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -911,12 +992,36 @@ def plot(self, filename=None):
         ax.spines['top'].set_color('none')
+        stepsize = 200
+        start, end = ax.get_xlim()
+        ax.xaxis.set_ticks(np.arange(start, end+1, stepsize))
```
This is dangerous. It will work for continental and global scale waveforms but not local ones where the whole waveform is only a couple of seconds.

If you want to tune the ticks please use one of matplotlibs tick locators. Otherwise I think this can also be reverted as matplotlib's default ticks are usually quite ok.

wjlei1990 replied:
Yes, it is not good. Let me remove it first and make some tests... I think I must do that for a purpose. Can not remember this time...

wjlei1990 replied:
See the picture I just made.
![aak ii 00 lhz](https://cloud.githubusercontent.com/assets/5777547/10611461/3a8e8070-771a-11e5-9c2d-dfcbcc0b11f8.png)
I think I remember why I did that... The axis increment is 1000 which I think is too large...(hard to figure out time on x axis)

The Green thick line in between is the selection time box(the selection mode is "body_wave" so it starts from first arrival till surface wave arrival). 

I will update it and leave it to you to modify since you are more experienced...

krischer replied:
> I will update it and leave it to you to modify since you are more experienced...

Haha...I just like to play around. Yea you can leave that to me if you want.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -911,12 +992,36 @@ def plot(self, filename=None):
+        #print(signal_start, signal_end)
+        text = "Signal Zone(s): [%-6.1f, %6.1f]" % (signal_start, signal_end)
+        plt.text(0.75, 0.81, text, horizontalalignment='left',
+                 verticalalignment='top', transform=ax.transAxes)
```
This could also be visually plotted which should make it even clearer. Maybe a gray background or a thick line along the data?

krischer replied:
Additionally it would also be great to see the time limits imposed by the various travel time limitations. `all_waves`, `body_waves`, ... then it is visually simpler to judge what is going on.

wjlei1990 replied:
Good idea! I am going to implement that...

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -911,12 +992,36 @@ def plot(self, filename=None):
+                 verticalalignment='top', transform=ax.transAxes)
+        dist_in_m, az, baz = geodetics.gps2DistAzimuth(self.event.latitude, self.event.longitude,
+                                self.station.latitude, self.station.longitude)
+        dist_in_degree = dist_in_m / (1000.0 * 40075.) * 360.0
```
http://docs.obspy.org/packages/autogen/obspy.core.util.geodetics.base.locations2degrees.html#obspy.core.util.geodetics.base.locations2degrees

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -911,12 +992,36 @@ def plot(self, filename=None):
+        plt.text(0.75, 0.99, text, horizontalalignment='left',
+                 verticalalignment='top', transform=ax.transAxes)
+        dist_in_m, az, baz = geodetics.gps2DistAzimuth(self.event.latitude, self.event.longitude,
+                                self.station.latitude, self.station.longitude)
```
Please check first if both event and station are given.

krischer commented:
So I looked through the code and commented on everything. Looks good! I really like the ability to select only certain waves.

The code has to be polished a bit before I can merge this. The most important things are that the tests have to pass and that the picked windows don't significantly change for existing users of pyflex.

Please also document all the changes in the various functions' docstrings and in the actual documentation if appropriate: https://github.com/krischer/pyflex/tree/master/doc

wjlei1990 commented:
Hi Lion,

Do I need to modify the documentation? It seems the pandoc is not installed on my desktop. If so, I also need to install pandoc to run "make html"...

I think I already modified the docstring in python files.

krischer commented:
The documentation actually gets build automatically each time a new commit is pushed to the master branch so you don't strictly need to build it. It might be useful if you change a lot so you can see how it looks in a rendered version.

Please make sure to run `make html` before you commit and changes to the `index.ipynb` - otherwise the built version will get commited and you'd have to rebase.

What OS do you have. On OSX pandoc can be installed with homebrew, on linux by the normal package manager.

wjlei1990 commented:
## Notes:

#### 1. Change all c_\* parameters to arrays

So people can set time-dependent parameters.
**Status: unresolved**

#### 2. Checking index

That would be good. We should probably just check that both indices are positive and the end index is larger then the start index. I think its reasonable to not allow negative indices although it would be valid in Python.
**Status: solved**

#### 3. Copy config object in WindowSelector

So it can be reused.
**Status: unresolved**

#### 4. Window SNR

One more thing, for the method "reject_based_on_signal_to_noise_ratio", I feel like we should provide people with one more option that uses both criteria, "amplitude" and "energy". At least for my case, I would like to have both checked... People can also chose which mode they want to use based on the features of their data. 
**Status: solved**

#### 5. Improve reject_on_select_mode()

I think it could also be expanded to select windows for specific phase arrivals. TauP does already calculate the arrival times for various phases so this function could well filter windows that contain the chosen phases.
**Status: unresolved**

#### 6. Curtail window difference(with original FLEXWIN).

"I first compared yours with the original fortran version and I found the difference. I feel like the original fortran version makes more sense. The curtailing left is just to curtail from the first left maxim. Same for the right maxim. It doesn't matter if the first left maxim is same as the center or the right maxium is same as the center. If in one window there is only one maxim(which is the center), it is still OK to curtail that window if the left or the right are too long. If you checked the FLEXWIN manual, in the curtail window example, the example window just contains one maxim and it is curtailed." 
**Status: solved**

#### 7. Polish plot.

Now we have noise region, signal region and selection region. We want to visualize them on the plots.
**Status: unresolved**

#### 8. Keep both observed.id and synthetic.id in window(?)

Do you think it is worth to keep both in window object. So for example, if I have two asdf file, one for observed and one for synthetic, and one window file. I can read in the observed.id and synthetic.id so I can know the window is selected on which two seismograms.

wjlei1990 commented:
Hi Lion,

Currently, I have two test can not be passed.

Btw, I modified the windows benchmarks in the test_pyflex.py since now the curtail window kicks in more frequently. But the window are basically the same.

## 1. code formatting

Did you see the same thing when you are running test_code_formatting? Or is it ok to leave them there. Cause I haven't touched these lines...

```
tests/test_code_formatting.py:55: AssertionError
---------------------------------------------------------- Captured stdout call -----------------------------------------------------------
/u/lei/software/wenjie/pyflex/src/pyflex/utils.py:19:1: E402 module level import not at top of file
/u/lei/software/wenjie/pyflex/src/pyflex/utils.py:20:1: E402 module level import not at top of file
/u/lei/software/wenjie/pyflex/src/pyflex/tests/test_pyflex.py:180:18: E712 comparison to False should be 'if cond is False:' or 'if not cond:'
/u/lei/software/wenjie/pyflex/src/pyflex/window_selector.py:19:1: E402 module level import not at top of file
/u/lei/software/wenjie/pyflex/src/pyflex/window_selector.py:20:1: E402 module level import not at top of file
/u/lei/software/wenjie/pyflex/src/pyflex/window_selector.py:21:1: E402 module level import not at top of file
/u/lei/software/wenjie/pyflex/src/pyflex/window_selector.py:22:1: E402 module level import not at top of file
/u/lei/software/wenjie/pyflex/src/pyflex/window_selector.py:23:1: E402 module level import not at top of file
/u/lei/software/wenjie/pyflex/src/pyflex/window_selector.py:24:1: E402 module level import not at top of file
/u/lei/software/wenjie/pyflex/src/pyflex/window_selector.py:25:1: E402 module level import not at top of file
/u/lei/software/wenjie/pyflex/src/pyflex/window_selector.py:26:1: E402 module level import not at top of file
/u/lei/software/wenjie/pyflex/src/pyflex/window_selector.py:27:1: E402 module level import not at top of file
/u/lei/software/wenjie/pyflex/src/pyflex/window_selector.py:29:1: E402 module level import not at top of file
/u/lei/software/wenjie/pyflex/src/pyflex/window_selector.py:30:1: E402 module level import not at top of file
/u/lei/software/wenjie/pyflex/src/pyflex/window_selector.py:31:1: E402 module level import not at top of file
/u/lei/software/wenjie/pyflex/src/pyflex/window_selector.py:32:1: E402 module level import not at top of file
```

## 2. test_window_plotting.

I don't want to replace the benchmark picture yet since I think you will(definitely) modify the plotting method in the future. Thus, to make the git repo clean and light, I don't want to replace the window figure. Once we are done with the plot method, we can replace the figure.

krischer commented:
@wjlei1990 The code formatting is because you used a newer flake8 version that has some more checks than what pyflex was originally developed against. But I fixed these like a month ago - I guess your branch started from an older version. If you rebase your branch on top of the current master branch these should disappear.

wjlei1990 committed:
fix code formatting issue

Two tests are still not passed:
1) code formatting: going to rebase to fix this problem
2) figure for tests should be replaced in the future

wjlei1990 committed:
Merge remote-tracking branch 'upstream/master' into devel

Conflicts:
	src/pyflex/tests/test_pyflex.py
	src/pyflex/window_selector.py

wjlei1990 commented:
@krischer I have merged devel with your master. So the code format issue is fixed. (I failed used "rebase" since there are patch failed issues...)

So issues left:
1). test_window_plotting: stated before
2) documentation: I added docstring but haven't touched anything under ./doc.
3) pytest for function I added in. You mentioned before, the "write_text()" in window_selector.py. How should I write the test? Compare two txt file to see if they are the same?

wjlei1990 committed:
reset some values to default ones

krischer commented on src/pyflex/config.py:
```diff
@@ -23,8 +23,10 @@ class Config(object):
                  dlna_acceptance_level=1.3, dlna_reference=0.0,
-                 cc_acceptance_level=0.7, s2n_limit=1.5, earth_model="ak135",
+                 cc_acceptance_level=0.7, earth_model="ak135",
+                 s2n_limit_energy=2.0, s2n_limit_amplitude=1.5,
```
This is tricky because it will unexpectedly break existing code for people who use the old `s2n_limit`. After this PR is released I'll release a new major version to make it clear that something changed.

Can you please also add the following:

``` python
def __init__(self, ..., **kwargs):
    if "s2n_limit" in kwargs:
        raise PyflexError(
            "The `s2n_limit` argument no longer exists. "
            "Please use `s2n_limit_energy` and/or `s2n_limit_amplitude`")
```

Then people will be aware that something changed and it raises a descriptive error with a clear message. I don't think we have to do a proper deprecation process for pyflex. Its too much work...

wjlei1990 replied:
What about keep the old name "s2n_limit" and add the new variable "s2n_limit_energy". So people don't need to change there existing scripts?

krischer replied:
Even better I think :-)

krischer commented on src/pyflex/config.py:
```diff
@@ -108,21 +111,39 @@ def __init__(self, min_period, max_period, stalta_waterlevel=0.07,
+            region. All windows containing data earlier than the min velocity
+            and later than the max velocity will be treated as surface waves.
+            Only used if station and event information is available.
         :type min_surface_wave_velocity: float
```
Change to `max_surface_wave_velocity`.

krischer commented on src/pyflex/config.py:
```diff
@@ -215,6 +236,19 @@ def __init__(self, min_period, max_period, stalta_waterlevel=0.07,
+            between min_surface_wave_velocity and max_surface_wave_velocity).
+            If ``"body_and_surface_wave"``, then pyflex will only accept
+            windows inside body and surface wave regions. If ``mantle_wave``,
+            pyflex will only accept windows after the surface wave region.
```
Lacks the `:type selection_mode: str` line.

Can you also please change the names to plural, e.g. to `body_waves`, `all_waves`, ... I think that is a bit nicer.

Really minor and its up to you if you want to do it. I will not insist on it: It might be nicer to document it an itemized list:

```
* ``"all_waves"``: All waves in the signal region.
* ``"body_waves"``: ...
```

krischer commented on src/pyflex/config.py:
```diff
@@ -247,15 +283,16 @@ def __init__(self, min_period, max_period, stalta_waterlevel=0.07,
-        if snr_type not in ["amplitude", "energy"]:
-            raise PyflexError("The window signal to noise type must be either"
-                              "'amplitude' or 'energy'.")
+        if snr_type not in ["amplitude", "energy", "amplitude_and_energy"]:
```
The new `"amplitude_and_energy"` possibility is not yet documented.

krischer commented on src/pyflex/window.py:
```diff
@@ -167,6 +167,21 @@ def relative_starttime(self):
     @property
+    def absolute_centertime(self):
+        """
+        Absolute time of the left border of this window.
```
Docstring is wrong.

krischer commented on src/pyflex/window.py:
```diff
@@ -167,6 +167,21 @@ def relative_starttime(self):
+    def relative_centertime(self):
+        """
+        Relative time of the left border in seconds to the first sample in
+        the array.
```
Docstring is wrong.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -186,6 +189,18 @@ def default(self, obj):
+    def write_text(self, filename):
+        """
+        Write plain text file for the window output
+        """
```
Can you add a short example to the docstring? Also please document the `filename` parameters.

wjlei1990 replied:
I have added documents. 

By "add a short example", you mean something like this?

```
    def write_text(self, filename):                                             
        """                                                                     
        Write windows to a plain text file. For example, after                  
        you selecting windows, you want to write them out and                   
        save as text files, you can call:                                       
        ws.write_text("window.txt")                                             

        :param filename: Name to write to.                                      
        :type filename: str                                                     
        """  
```

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -318,32 +343,37 @@ def select_windows(self):
         self.reject_windows_based_on_minimum_length()
         self.reject_based_on_signal_to_noise_ratio()
         self.reject_based_on_data_fit_criteria()
+        self.__print_remaining_windows()
```
We'll eventually have to benchmark these calls. They all call `logger.debug` for each and every window. Sometimes the algorithm results in a ridiculous number of temporary windows where all these function calls might actually might make a difference. But that is just something to keep in mind for now.

wjlei1990 replied:
By "might actually make a difference", you mean introduce bugs?

krischer replied:
No but it might influence performance. The idea is that `logger.debug` calls are not really pursued if the log level is lower than `DEBUG` so they almost free. But if they are called thousands upon thousands of times it might have an impact on performance. Just something to keep in mind and change later if necessary.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -402,12 +433,22 @@ def determine_signal_and_noise_indices(self):
             else:
                 self.config.noise_end_index = \
-                    int(self.ttimes[0]["time"] - self.config.min_period)
+                    int(self.ttimes[0]["time"] - 2.*self.config.min_period)
```
Wrong formatting. Just use `2 * self.con...`.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -402,12 +433,22 @@ def determine_signal_and_noise_indices(self):
+
+        dt = self.observed.stats.delta
+        logger.info("Noise region [%ds, %ds]; signal region [%ds, %ds]" % (
+                    self.config.noise_start_index*dt,
```
Also wrong formatting. There needs to be spaces around the asterix.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -421,31 +462,47 @@ def reject_based_on_signal_to_noise_ratio(self):
-        if self.config.window_signal_to_noise_type == "amplitude":
-            noise_amp = np.abs(noise).max()
+        def filter_window_noise_amplitude(win):
+            win_signal = self.observed.data[win.left: win.right]
```
no space around the colon.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -421,31 +462,47 @@ def reject_based_on_signal_to_noise_ratio(self):
-                    return False
-                return True
+        def filter_window_noise_energy(win):
+            data = self.observed.data[win.left: win.right]
```
no space around the colon.

krischer commented on src/pyflex/window_selector.py:
```diff
@@ -421,31 +462,47 @@ def reject_based_on_signal_to_noise_ratio(self):
+            self.windows = list(filter(filter_window_noise_energy,
+                                       self.windows))
+
+        elif window_snr_type == "amplitude_and_energy":
```
No need for this third condition. Just replace the upper two with:

``` python
if window_snr_type in ("amplitude", "amplitude_and_energy"):
    ...
elif window_snr_type in ("energy", "amplitude_and_energy"):
    ...
```

krischer commented:
I put some more comments. Can you also give me push rights to your fork? Then I can look at the plotting on monday.

Have a nice weekend!

wjlei1990 commented:
Hi Lion,

I have added you in my repo.

I will look at your comments and fix the issues later.

Best,

## Wenjie

 Graduate student in Geophysics
 Princeton University | Guyot Hall 309
 Phone: 609-937-4659

---

From: Lion Krischer [notifications@github.com]
Sent: Friday, November 13, 2015 13:14
To: krischer/pyflex
Cc: Wenjie Lei
Subject: Re: [pyflex] Devel (#7)

I put some more comments. Can you also give me push rights to your fork? Then I can look at the plotting on monday.

Have a nice weekend!

—
Reply to this email directly or view it on GitHubhttps://github.com/krischer/pyflex/pull/7#issuecomment-156506012.

wjlei1990 committed:
fix code formatting issues suggested by Lion

wjlei1990 committed:
fix s2n_limit naming issue

wjlei1990 committed:
close figure handle after plotting

wjlei1990 committed:
add checks for plotting

wjlei1990 committed:
fix bug in noise_end_index calculation and refine __eq__ in Window

wjlei1990 committed:
Merge branch 'devel' of https://github.com/wjlei1990/pyflex into devel

wjlei1990 committed:
Merge branch 'devel' of https://github.com/wjlei1990/pyflex into devel

wjlei1990 committed:
PEP8 formatting issue

wjlei1990 committed:
update baseline image to pass the test

wjlei1990 commented:
Hi Lion,

I just created a new pull request. I replaced the baseline image so all tests will pass this time.

But travis give this error: https://travis-ci.org/krischer/pyflex/jobs/118038274

It seems miniconda is not installed correctly.

wjlei1990 committed:
add channel_id_2(to keep track of synt trace id)

krischer commented:
Hey Wenjie,

probably time to merge this into the master ;-) Is this the latest version you have? I'd then have a more extensive look at it.

wjlei1990 committed:
remove the image test since it depands on the specific version of matplotlib(should be added back later)

wjlei1990 committed:
refactor and refine the code in selection mode and noise region rejection

1) modify `calculate_preliminary` to check the signal region instead
2) add `reject_on_noise_region`, split from `reject_on_selection_mode`
3) add selection_mode=`custom` for user customization

wjlei1990 committed:
add 'is None' test to avoid 0 value conditional trap

wjlei1990 committed:
add tests to check noise index

wjlei1990 committed:
change the obspy module import path according to updates in obspy

wjlei1990 committed:
change reject_on_selection_mode to window.center

wjlei1990 commented:
Sorry, missed this thread...Can we work together with the merge now on?
